### PR TITLE
Use rubocop -a ; rubocop --auto-gen-config.

### PR DESCRIPTION
**Pull request by dblock**
I expect lots of strong opinions and don't expect this to be merged, really ;)

This comes from https://github.com/artsy/bearden/pull/311#discussion_r143014691

> I think a .rubocop_todo.yml is an antipattern - why run RuboCop if you're just going to ignore what it says!!

Ruby style (eg. tabs vs. spaces) is a contentious topic, introducing Rubocop allows to _defer style to a tool_ when used in this manner.

1) Standardizing on `rubocop -a ; rubocop --auto-gen-config` is the least mental overhead, enforce only what can be autocorrected, use defaults provided by the tool.
2) Autocorrect does a great job at spaces vs. tabs, alignment and parenthesis and `raise` vs. `fail`, as this PR shows how much it was able to autocorrect vs. what was autocorrected without a hand-crafted configuration.
3) You rarely or ever have to think/edit what's in the `.yml` files, it's the least amount of work.
4) It's the least common denominator for the entire organization for 100+ repositories. If we handcraft things in `rubocop.yml` we end up with more than one standard.
5) Never put a `rubocop:disable/enable` for silly things like method lengths or comments, which are all over the place in this codebase.
6) Never use `rubocop:disable all` because we give up, like this project does in multiple places, providing an opportunity to examine violations in one place and actually address them.

### CODE REVIEW

**Comment by damassi** [comment]
This, I'm assuming, follows the same philosophy as [Prettier](https://github.com/prettier/prettier)?

**Comment by dblock** [comment]
Indeed @damassi, the goal, just like with Prettier, is to stop all debates about style.

**Comment by orta** [comment]
Reader: You may be confused at what a rubocop todo is, I was. [Here's one](https://github.com/puppetlabs/puppetlabs-apt/blob/master/.rubocop_todo.yml#L1) - it's a list of violations from running auto-correct on the entire project. I assume that this also strips the existing `//disables`? 

I'm a more middle of the roads here, I'm a 👍 on rubocop in general so long as it autocorrects whatever people have specific differing opinions about. E.g. if I'm gonna own/create an artsy repo in ruby, I have a pretty strong opinion on double quotes vs single ;) - I'd expect rubocop to switch `'` to `"` for people automatically. Which is what we do in JS repos.

I like the idea of the TODO list, seems like a nice abstraction. Hrm, not really sure I have much to add here actually.

**Comment by jonallured** [comment]
RuboCop is not just a linter, it also analyzes code. From the RuboCop README:

> Apart from reporting problems in your code, RuboCop can also automatically fix some of the problems for you.

My problem with running `rubocop -a ; rubocop --auto-gen-config` is that "some" word up there. :wink:

What types of violations can RuboCop correct? Let's play.

```ruby
# test-script.rb
omg = "asdfasdf"
```

```
jon@magneto:~/code/lab_notes(master+)% rubocop test-script.rb -a
Inspecting 1 file
W

Offenses:

test-script.rb:1:1: W: Useless assignment to variable - omg.
omg = 'asdfasdf'
^^^
test-script.rb:1:1: C: The name of this source file (test-script.rb) should use snake_case.
omg = 'asdfasdf'
^
test-script.rb:1:7: C: [Corrected] Prefer single-quoted strings when you don't need string interpolation or special symbols.
omg = "asdfasdf"
      ^^^^^^^^^^

1 file inspected, 3 offenses detected, 1 offense corrected
```

So it's able to fix the type of quote used, but it can't do anything about the useless assignment or the weird filename picked. Had I added the `--auto-gen-config` flag, RuboCop would have added exceptions so that those violations would be ignored, the build would be green and this code would be easy to merge.

This is what I dislike - I am happy to leave quotes to linters, but violations that can't be auto-corrected should not be ignored by default. Useless assignment or a filename I can probably live with, but there are way worse offenses that would be ignored, probably the most important (to me) are:

* [Metrics/AbcSize](http://rubocop.readthedocs.io/en/latest/cops_metrics/#metricsabcsize)
* [Metrics/CyclomaticComplexity](http://rubocop.readthedocs.io/en/latest/cops_metrics/#metricscyclomaticcomplexity)
* [Metrics/PerceivedComplexity](http://rubocop.readthedocs.io/en/latest/cops_metrics/#metricsperceivedcomplexity)

When code violates these cops and instead of refactoring, exceptions are added, that's a red flag for me. And this is how prettier and RuboCop are different - the analysis part. Standard does some of this, but I don't know if it does the code complexity stuff that RuboCop can do.

And none of this is to say that RuboCop always gets it right, but I would prefer for a human to have to make this determination. If you write a method that can't be refactored to comply with these complexity metrics, then you can disable the cop for that method and justify yourself in your PR message. Then the review process will either confirm or suggest other ways of approaching your problem.

---

I agree with most of your first 3 points - let's look at the rest:

> 4. It's the least common denominator for the entire organization for 100+ repositories. If we handcraft things in rubocop.yml we end up with more than one standard.

Yeah, I see what you mean on this one, however, if you look at the `rubocop_todo.yml` files across the org, you'll find lots of places where we've written complex code that's hard to maintain and added an exception. That's not good either. I'm looking at you Volt.

> 5. Never put a rubocop:disable/enable for silly things like method lengths or comments, which are all over the place in this codebase.

I'd love to hear more about why you feel this way.

> 6. Never use rubocop:disable all because we give up, like this project does in multiple places, providing an opportunity to examine violations in one place and actually address them.

The `data-exploration` folder where those can all be found was throw-away code that Lance wanted to check in, but didn't bother to fix up. It can probably be removed at this point. I don't think we've given up anywhere in this codebase and in fact, I've had to examine many parts in order to comply with RuboCop to the betterment of the codebase!

I've worked really hard to follow RuboCop defaults and only ignore when I felt it was warranted. Some of this is preference, but I feel pretty strongly about complexity violations - why encourage a practice that lets devs who have written complex code off the hook so easily?

**Comment by orta** [comment]
> RuboCop is not just a linter, it also analyzes code

Yeah, I think this is why prettier was created, leave code semantic analysis to other projects and let style be 100% automated away.

**Comment by jonallured** [comment]
Sorry, one more bit of follow up here - I was curious to get some data around this topic, turns out you can do this:

```ruby
> total_cops = RuboCop::Cop::Cop.registry.cops
> autocorrecting_cops = RuboCop::Cop::Cop.registry.cops.select { |c| c.new.support_autocorrect? }
> total_cops.count - autocorrecting_cops.count
=> 123
```

For those that are interested, here are the cops that are ignored when you use the approach of ignoring violations that can't be autocorrected:

<details>

* RuboCop::Cop::Bundler::DuplicatedGem
* RuboCop::Cop::Layout::EndOfLine
* RuboCop::Cop::Lint::AmbiguousBlockAssociation
* RuboCop::Cop::Lint::AmbiguousOperator
* RuboCop::Cop::Lint::AmbiguousRegexpLiteral
* RuboCop::Cop::Lint::AssignmentInCondition
* RuboCop::Cop::Lint::BooleanSymbol
* RuboCop::Cop::Lint::CircularArgumentReference
* RuboCop::Cop::Lint::ConditionPosition
* RuboCop::Cop::Lint::Debugger
* RuboCop::Cop::Lint::DuplicateCaseCondition
* RuboCop::Cop::Lint::DuplicateMethods
* RuboCop::Cop::Lint::DuplicatedKey
* RuboCop::Cop::Lint::EachWithObjectArgument
* RuboCop::Cop::Lint::ElseLayout
* RuboCop::Cop::Lint::EmptyExpression
* RuboCop::Cop::Lint::EmptyWhen
* RuboCop::Cop::Lint::EndInMethod
* RuboCop::Cop::Lint::EnsureReturn
* RuboCop::Cop::Lint::FloatOutOfRange
* RuboCop::Cop::Lint::FormatParameterMismatch
* RuboCop::Cop::Lint::HandleExceptions
* RuboCop::Cop::Lint::ImplicitStringConcatenation
* RuboCop::Cop::Lint::IneffectiveAccessModifier
* RuboCop::Cop::Lint::InterpolationCheck
* RuboCop::Cop::Lint::InvalidCharacterLiteral
* RuboCop::Cop::Lint::LiteralInCondition
* RuboCop::Cop::Lint::Loop
* RuboCop::Cop::Lint::NestedMethodDefinition
* RuboCop::Cop::Lint::NextWithoutAccumulator
* RuboCop::Cop::Lint::NonLocalExitFromIterator
* RuboCop::Cop::Lint::ParenthesesAsGroupedExpression
* RuboCop::Cop::Lint::RandOne
* RuboCop::Cop::Lint::RequireParentheses
* RuboCop::Cop::Lint::RescueException
* RuboCop::Cop::Lint::RescueWithoutErrorClass
* RuboCop::Cop::Lint::ReturnInVoidContext
* RuboCop::Cop::Lint::ShadowedException
* RuboCop::Cop::Lint::ShadowingOuterLocalVariable
* RuboCop::Cop::Lint::Syntax
* RuboCop::Cop::Lint::UnderscorePrefixedVariableName
* RuboCop::Cop::Lint::UnreachableCode
* RuboCop::Cop::Lint::UriEscapeUnescape
* RuboCop::Cop::Lint::UselessAccessModifier
* RuboCop::Cop::Lint::UselessAssignment
* RuboCop::Cop::Lint::UselessComparison
* RuboCop::Cop::Lint::UselessElseWithoutRescue
* RuboCop::Cop::Lint::UselessSetterCall
* RuboCop::Cop::Lint::Void
* RuboCop::Cop::Metrics::CyclomaticComplexity
* RuboCop::Cop::Metrics::AbcSize
* RuboCop::Cop::Metrics::BlockLength
* RuboCop::Cop::Metrics::BlockNesting
* RuboCop::Cop::Metrics::ClassLength
* RuboCop::Cop::Metrics::LineLength
* RuboCop::Cop::Metrics::MethodLength
* RuboCop::Cop::Metrics::ModuleLength
* RuboCop::Cop::Metrics::ParameterLists
* RuboCop::Cop::Metrics::PerceivedComplexity
* RuboCop::Cop::Naming::AccessorMethodName
* RuboCop::Cop::Naming::AsciiIdentifiers
* RuboCop::Cop::Naming::ClassAndModuleCamelCase
* RuboCop::Cop::Naming::ConstantName
* RuboCop::Cop::Naming::FileName
* RuboCop::Cop::Naming::HeredocDelimiterCase
* RuboCop::Cop::Naming::HeredocDelimiterNaming
* RuboCop::Cop::Naming::MethodName
* RuboCop::Cop::Naming::BinaryOperatorParameterName
* RuboCop::Cop::Naming::PredicateName
* RuboCop::Cop::Naming::VariableName
* RuboCop::Cop::Naming::VariableNumber
* RuboCop::Cop::Performance::Caller
* RuboCop::Cop::Performance::FixedSize
* RuboCop::Cop::Performance::UnfreezeString
* RuboCop::Cop::Style::AsciiComments
* RuboCop::Cop::Style::AutoResourceCleanup
* RuboCop::Cop::Style::BeginBlock
* RuboCop::Cop::Style::CaseEquality
* RuboCop::Cop::Style::ClassAndModuleChildren
* RuboCop::Cop::Style::ClassVars
* RuboCop::Cop::Style::DocumentationMethod
* RuboCop::Cop::Style::Documentation
* RuboCop::Cop::Style::DoubleNegation
* RuboCop::Cop::Style::EndBlock
* RuboCop::Cop::Style::FlipFlop
* RuboCop::Cop::Style::For
* RuboCop::Cop::Style::FormatStringToken
* RuboCop::Cop::Style::GlobalVars
* RuboCop::Cop::Style::GuardClause
* RuboCop::Cop::Style::IdenticalConditionalBranches
* RuboCop::Cop::Style::IfInsideElse
* RuboCop::Cop::Style::IfUnlessModifierOfIfUnless
* RuboCop::Cop::Style::IfWithSemicolon
* RuboCop::Cop::Style::ImplicitRuntimeError
* RuboCop::Cop::Style::InlineComment
* RuboCop::Cop::Style::MethodCalledOnDoEndBlock
* RuboCop::Cop::Style::MethodMissing
* RuboCop::Cop::Style::MissingElse
* RuboCop::Cop::Style::ModuleFunction
* RuboCop::Cop::Style::MultilineBlockChain
* RuboCop::Cop::Style::MultilineTernaryOperator
* RuboCop::Cop::Style::MultipleComparison
* RuboCop::Cop::Style::NestedTernaryOperator
* RuboCop::Cop::Style::OptionHash
* RuboCop::Cop::Style::OptionalArguments
* RuboCop::Cop::Style::Send
* RuboCop::Cop::Style::SingleLineBlockParams
* RuboCop::Cop::Style::StructInheritance
* RuboCop::Cop::Rails::Date
* RuboCop::Cop::Rails::EnumUniqueness
* RuboCop::Cop::Rails::Exit
* RuboCop::Cop::Rails::FilePath
* RuboCop::Cop::Rails::HasAndBelongsToMany
* RuboCop::Cop::Rails::HasManyOrHasOneDependent
* RuboCop::Cop::Rails::NotNullColumn
* RuboCop::Cop::Rails::OutputSafety
* RuboCop::Cop::Rails::Output
* RuboCop::Cop::Rails::ReversibleMigration
* RuboCop::Cop::Rails::ScopeArgs
* RuboCop::Cop::Rails::SkipsModelValidations
* RuboCop::Cop::Rails::TimeZone
* RuboCop::Cop::Security::Eval
* RuboCop::Cop::Security::MarshalLoad

</details>

**Comment by dblock** [comment]
> Never put a rubocop:disable/enable for silly things like method lengths or comments, which are all over the place in this codebase.
> I'd love to hear more about why you feel this way.

Because this is the equivalent of `rubocop --auto-gen-config` but hidden all over the place inside code, I think they are noise.

**Comment by dblock** [comment]
> I've worked really hard to follow RuboCop defaults and only ignore when I felt it was warranted. Some of this is preference, but I feel pretty strongly about complexity violations - why encourage a practice that lets devs who have written complex code off the hook so easily?

I've seen devs bend code to fit in the complexity space just like they would try to add a `\` at the end of a line to get rid of a "line too long" violation, making code unreadable.

**Comment by dblock** [comment]
I am all for improving the code btw, I often kill the `_todo` file, then run rubocop and try to fix the most blatant violations.

**Comment by dblock** [comment]
Finally, this is a discussion we should have across the company, much like API style, and have some guidelines, cc: @joeyAghion @orta.

**Comment by starsirius** [comment]
I think we are talking about different goals here:
1. **only enforce what can be auto-corrected and achieve the least mental overhead.**
  I feel what can be auto-corrected was still somewhat limited (tho looks like it will [expand](https://github.com/bbatsov/rubocop/wiki/Automatic-Corrections)). Running `rubocop --auto-correct ; rubocop --auto-gen-config` would likely still lead to pretty messy/non-standardized/methods-with-100-lines/etc code. I'm slightly concerned about that.
1. **respect Rubocop default styles/configs and be strict. Relax when necessary.**
  Running `rubocop --auto-correct` saved me a lot of time before. For things that cannot be auto-corrected, we will still need to do our diligence to refactor. My problem with `.rubocop_todo.yml` was that it's never a "todo". We rarely went back to remove items from them one by one, as suggested in the [manual](https://github.com/bbatsov/rubocop/blob/master/manual/configuration.md#automatically-generated-configuration). It became a lazy way to pass Rubocop.

Maybe we could talk about how much we want to standardize within the organization.

Other thoughts about @dblock's points:
> 4. It's the least common denominator for the entire organization for 100+ repositories. If we handcraft things in rubocop.yml we end up with more than one standard.

Using `rubocop --auto-correct ; rubocop --auto-gen-config` across the org means only standardizing minimal styles, and in individual projects we might end up with either still more than one standard or no standard for the majority of stuff. Maybe another way without overly handcrafting rubocop.yml rules in individual project is to come up with a org-level rubocop.yml [gem](https://github.com/bbatsov/rubocop/blob/master/manual/configuration.md#inheriting-configuration-from-a-dependency-gem) that projects can inherit from (and customize if necessary).

> 5. Never put a rubocop:disable/enable for silly things like method lengths or comments, which are all over the place in this codebase.
> 6. Never use rubocop:disable all because we give up, like this project does in multiple places, providing an opportunity to examine violations in one place and actually address them.
I agreed with these. We should avoid code-level rules when possible.

**Comment by joeyAghion** [comment]
I feel similarly in that I've strived to _personally_ abide by most of Rubocop's default cops, but hesitate to add it to projects because
* the cops themselves frequently change, necessitating noisy code-thrashing
* there's little distinction between basic style enforcement and quality critiques

I could see us getting good use out of a radically stripped-down, shared `.rubocop.yml` that declares our actual style expectations and leaves out some of the length/complexity cops, but this seems to fight against the gem's nature. I'd also like to avoid having to write code _about rubocop_ (such as enabling/disabling comments) at all costs.

In summary, I think it would be nice to _enforce_ a small set of rules but _warn_ developers about a larger set of antipatterns/smells.

**Comment by izakp** [comment]
My two cents here is to treat Rubocop as a simply linter, use a bare minimal set of rules, and leave the refactoring to humans (agreeing here with @jonallured).  I personally find rules [like this](https://github.com/artsy/gravity/commit/c8ee3072c64ab003309af4acdbce40e24287806e) very pedantic and aside from the very basic single vs double quotes / indentation / missing variable assignment rules I am generally not in favor of Rubocop's suggestions.  When it comes to style I generally believe less is more.

In general I don't think we should go too far [down the rabbit-hole](https://www.theregister.co.uk/2016/07/11/linus_torvalds_in_sweary_rant_about_punctuation_in_kernel_comments/) of prescriptive style, and disagree here with the approach used by Prettier, as @dblock says "stop all debates about style" or as @orta puts it "let style be 100% automated away" because code style is relatively subjective human problem and as such I believe we should treat it relatively lightly, by discussion and by means of humans (i.e. devs in dialogue with other devs).  Trying to find a one-size-fits-all solution this perceived problem is IMHO overengineering the actual process of development, which is in the end a process of writing languages designed for and built by humans as an interface to machine language.

**Comment by damassi** [comment]
One of the least exciting things I do during the work day is fight with Prettier. A good eslint setup with its sensible defaults that catch basic errors while leaving me to my own code-writing process will win over it any day.
